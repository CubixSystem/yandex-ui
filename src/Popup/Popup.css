/**
 * Без important стили класса перебиваются inline-стилями
 */
.Popup2 {
    position: absolute;
    z-index: 1400;
}

.Popup2_outside {
    top: -10000px !important;
    left: -10000px !important;

    display: block !important;
}
