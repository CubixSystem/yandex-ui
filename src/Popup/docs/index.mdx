---
id: Popup
path: /components/popup

title: Popup
description: Компонент для создания всплывающего окна (попапа).

links:
 - label: source
   url: https://github.yandex-team.ru/search-interfaces/frontend/tree/master/packages/lego-components/src/Popup
 - label: startrek
   url: https://search.yandex-team.ru/stsearch?text=Popup.ts&facet.queue=ISL&facet.type=bug&facet.status=128

tabs:
 - Examples
 - Usage
 - Playground
 - Props
---