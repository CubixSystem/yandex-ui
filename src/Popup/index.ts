export * from './Popup';
export * from './_nonvisual/Popup_nonvisual';
export * from './_target/Popup_target_anchor';
export * from './_theme/Popup_theme_clear';
export * from './_theme/Popup_theme_normal';
export * from './_view/Popup_view_default';
export { Direction, directions } from '../usePopper';
