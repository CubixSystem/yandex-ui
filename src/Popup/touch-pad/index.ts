export * from '../Popup@touch-pad';
export * from '../_nonvisual/Popup_nonvisual';
export * from '../_target/Popup_target_anchor';
export * from '../_theme/Popup_theme_clear';
export * from '../_theme/Popup_theme_normal';
export * from '../_view/Popup_view_default';
