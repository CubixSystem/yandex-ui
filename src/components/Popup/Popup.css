/**
 * Без important стили класса перебиваются inline-стилями
 */
.Popup2 {
    position: absolute;
}

.Popup2_outside {
    top: -10000px !important;
    left: -10000px !important;

    display: block !important;
}

.Popup2.Popup2_visible {
    z-index: 1;
}
