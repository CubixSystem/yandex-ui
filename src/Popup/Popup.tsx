import React, {
    RefObject,
    ReactNode,
    FC,
    CSSProperties,
    useEffect,
    useState,
    ReactElement,
    Ref,
    useRef,
    useCallback,
    MouseEventHandler,
} from 'react';
import { createPortal } from 'react-dom';
import { useComponentRegistry } from '@bem-react/di';
import { cn } from '@bem-react/classname';

import { canUseDOM } from '../lib/canUseDOM';
import { mergeAllRefs } from '../lib/mergeRefs';
import { IPopupRegistry } from './Popup.registry';

import './Popup.css';
import { useUpdateEffect } from '../useUpdateEffect';
import { OnClose, LayerManager } from '../LayerManager/LayerManager';

export type Direction =
    | 'bottom-left'
    | 'bottom-center'
    | 'bottom-right'
    | 'top-left'
    | 'top-center'
    | 'top-right'
    | 'right-top'
    | 'right-center'
    | 'right-bottom'
    | 'left-top'
    | 'left-center'
    | 'left-bottom';

export type Position = {
    top: number;
    left: number;
};

export type OptionalPosition = {
    top?: number;
    left?: number;
    bottom?: number;
    right?: number;
};

export type DrawingParams = {
    direction: Direction;
    height: number;
    left: number;
    top: number;
    width: number;
};

export interface IPopupProps {
    /**
     * Дополнительный контент после содержимого попапа
     */
    addonAfter?: ReactNode;

    /**
     * Дополнительный контент перед содержимым попапа
     */
    addonBefore?: ReactNode;

    /**
     * Задает направление хвостика. Например, если указано значение `bottom-center` — хвостик выходит из центра снизу.
     *
     * Свойство `direction` необходимо использовать без модификатора `target_anchor`.
     * Чтобы задать направление раскрытия для попапа с модификатором `target_anchor`,
     * установите свойство `directions`
     */
    direction?: Direction;

    /**
     * Вызывает дополнительный рендер после создания
     * @deprecated Используйте `visible`
     */
    forceRender?: boolean;

    /**
     * Включает/отключает хвостик у попапа
     */
    hasTail?: boolean;

    /**
     * Ссылка на корневой DOM-элемент компонента
     */
    innerRef?: Ref<HTMLDivElement>;

    /**
     * Сохраняет контейнер в DOM после создания
     *
     * @default true
     */
    keepMounted?: boolean;

    /**
     * Задает позицию попапа. Свойство `position` необходимо использовать без модификатора `target_anchor`
     */
    position?: OptionalPosition;

    /**
     * Ссылка на DOM-элемент, в котором размещается попап
     *
     * Важно, чтобы контейнер имел `position: relative` для корректного позициоинирования.
     *
     * @default document.body
     */
    scope?: RefObject<HTMLElement>;

    /**
     * Задает позицию хвостика. Свойство `tailPosition` необходимо использовать без модификатора `target_anchor`.
     */
    tailPosition?: Position;

    /**
     * Ссылка на DOM-элемент хвостика
     */
    tailRef?: Ref<HTMLDivElement>;

    /**
     * Задает размер хвостика
     */
    tailSize?: number;

    /**
     * Делает попап видимым
     */
    visible?: boolean;

    /**
     * Задает слой `z-index`
     */
    zIndex?: number;

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Html атрибут `style`
     */
    style?: CSSProperties;

    /**
     * Функция, вызывающаяся при отрисовке хвостика.
     * Вызывается вне зависимости от наличия флага `hasTail`.
     */
    unstable_onRenderTail?: (tail: ReactElement) => ReactElement;

    /**
     * Содержимое попапа
     */
    children?: ReactNode | ((props: { tailRef?: Ref<HTMLDivElement> }) => ReactNode);

    /**
     * Обработчик, вызывающийся после нажатия на клавишу esc либо мышкой на область вне контейнера
     */
    onClose?: OnClose;

    /**
     * Список ссылок на DOM-узлы в рамках которых не нужно отслеживать нажатия
     *
     * @internal
     */
    unstable_essentialRefs?: RefObject<HTMLElement>[];

    /**
     * DOM-узел в рамках которого не нужно отслеживать нажатие
     *
     * @internal
     */
    unstable_hostRef?: RefObject<HTMLElement>;

    /**
     * Обработчик, вызываемый при срабатывании события click
     */
    onClick?: MouseEventHandler<HTMLDivElement>;
}

type PopupInternalProps = IPopupProps & {
    onOutsideClick?: (event: any) => void;
    onEscapeKeyDown?: (event: any) => void;
};

export const cnPopup = cn('Popup2');

/**
 * Компонент для создания всплывающего окна (попапа).
 * @param {IPopupProps} props
 */
export const Popup: FC<IPopupProps> = ({
    addonAfter,
    addonBefore,
    children,
    className,
    direction,
    forceRender,
    hasTail,
    innerRef,
    keepMounted = true,
    position,
    scope = { current: canUseDOM() ? document.body : null },
    style,
    tailPosition,
    tailRef,
    tailSize,
    visible,
    zIndex,
    unstable_onRenderTail,
    onClose,
    unstable_essentialRefs = [],
    unstable_hostRef,
    // TODO: https://st.yandex-team.ru/ISL-8670
    onOutsideClick,
    onEscapeKeyDown,
    onClick,
    // Извлекаем свойства, т.к. они не нужны на DOM узле
    // FIXME: https://github.com/bem/bem-react/issues/381
    // @ts-ignore
    theme: _theme,
    // @ts-ignore
    view: _view,
    // @ts-ignore
    nonvisual: _nonvisual,
    ...props
}: PopupInternalProps) => {
    const [isFirstRender, forceUpdate] = useState(true);
    const containerRef = useRef(null);

    useUpdateEffect(() => {
        if (isFirstRender && visible) {
            forceUpdate(false);
        }
    }, []);

    useEffect(() => {
        console.assert(
            onOutsideClick === undefined || onEscapeKeyDown === undefined,
            'Использование функции "withOutsideClick" является устаревшим API. ' +
                'Для закрытия используйте свойство "onClose" без использования "withOutsideClick".',
        );

        if (isFirstRender && (forceRender || visible)) {
            forceUpdate(false);
        }
    }, []);

    const onCloseRequest = useCallback((event, source) => {
        if (source === 'esc') {
            if (onEscapeKeyDown !== undefined) {
                onEscapeKeyDown(event);
            }
        } else if (onOutsideClick !== undefined) {
            onOutsideClick(event);
        }
        if (onClose !== undefined) {
            onClose(event, source);
        }
    }, [onEscapeKeyDown, onOutsideClick, onClose]);

    const { Tail } = useComponentRegistry<IPopupRegistry>(cnPopup());
    const hostRef = unstable_hostRef || containerRef;

    if ((!visible && !keepMounted) || !canUseDOM() || scope.current === null || isFirstRender) {
        return null;
    }

    return createPortal(
        <LayerManager visible={visible} onClose={onCloseRequest} essentialRefs={[hostRef, ...unstable_essentialRefs]}>
            <div
                {...props}
                className={cnPopup({ visible, direction }, [className])}
                ref={mergeAllRefs(containerRef, innerRef)}
                style={{ ...style, ...position, zIndex }}
                onClick={onClick}
            >
                {addonBefore}
                {typeof children === 'function' ? children({ tailRef }) : children}
                {addonAfter}
                {unstable_onRenderTail &&
                    unstable_onRenderTail(
                        <Tail innerRef={tailRef} style={{ ...tailPosition, height: tailSize, width: tailSize }} />,
                    )}
                {!unstable_onRenderTail && hasTail && (
                    <Tail innerRef={tailRef} style={{ ...tailPosition, height: tailSize, width: tailSize }} />
                )}
            </div>
        </LayerManager>,
        scope.current,
    );
};

Popup.displayName = cnPopup();
