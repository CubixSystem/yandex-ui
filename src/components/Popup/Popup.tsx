import React, { RefObject, ReactNode, FC, CSSProperties, useEffect, useState, ReactElement, Ref } from 'react';
import { createPortal } from 'react-dom';
import { useComponentRegistry } from '@bem-react/di';
import { cn } from '@bem-react/classname';

import { canUseDOM } from '../../lib/canUseDOM';
import { mergeAllRefs } from '../../lib/mergeRefs';
import { IPopupRegistry } from './Popup.registry';

import './Popup.css';
import { useUpdateEffect } from '../../hooks/useUpdateEffect';

export type Direction =
    | 'bottom-left'
    | 'bottom-center'
    | 'bottom-right'
    | 'top-left'
    | 'top-center'
    | 'top-right'
    | 'right-top'
    | 'right-center'
    | 'right-bottom'
    | 'left-top'
    | 'left-center'
    | 'left-bottom';

export type Position = {
    top: number;
    left: number;
};

export type OptionalPosition = {
    top?: number;
    left?: number;
    bottom?: number;
    right?: number;
};

export type DrawingParams = {
    direction: Direction;
    height: number;
    left: number;
    top: number;
    width: number;
};

export interface IPopupProps {
    /**
     * Дополнительный контент после содержимого попапа
     */
    addonAfter?: ReactNode;

    /**
     * Дополнительный контент перед содержимым попапа
     */
    addonBefore?: ReactNode;

    /**
     * Задает направление хвостика. Например, если указано значение `bottom-center` — хвостик выходит из центра снизу.
     *
     * Свойство `direction` необходимо использовать без модификатора `target_anchor`.
     * Чтобы задать направление раскрытия для попапа с модификатором `target_anchor`,
     * установите свойство `directions`
     */
    direction?: Direction;

    /**
     * Вызывает дополнительный рендер после создания
     * @deprecated Используйте `visible`
     */
    forceRender?: boolean;

    /**
     * Включает/отключает хвостик у попапа
     */
    hasTail?: boolean;

    /**
     * Ссылка на корневой DOM-элемент компонента
     */
    innerRef?: Ref<HTMLDivElement>;

    /**
     * Сохраняет контейнер в DOM после создания
     *
     * @default true
     */
    keepMounted?: boolean;

    /**
     * Задает позицию попапа. Свойство `position` необходимо использовать без модификатора `target_anchor`
     */
    position?: OptionalPosition;

    /**
     * Ссылка на DOM-элемент, в котором размещается попап
     *
     * Важно, чтобы контейнер имел `position: relative` для корректного позициоинирования.
     *
     * @default document.body
     */
    scope?: RefObject<HTMLElement>;

    /**
     * Задает позицию хвостика. Свойство `tailPosition` необходимо использовать без модификатора `target_anchor`.
     */
    tailPosition?: Position;

    /**
     * Ссылка на DOM-элемент хвостика
     */
    tailRef?: Ref<HTMLDivElement>;

    /**
     * Задает размер хвостика
     */
    tailSize?: number;

    /**
     * Ссылка на DOM-элемент, в котором не отслеживаются нажатия. Используется с `withOutsideClick`
     */
    targetRef?: Ref<HTMLElement>;

    /**
     * Делает попап видимым
     */
    visible?: boolean;

    /**
     * Задает слой `z-index`
     */
    zIndex?: number;

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Html атрибут `style`
     */
    style?: CSSProperties;

    /**
     * Функция, вызывающаяся при отрисовке хвостика.
     * Вызывается вне зависимости от наличия флага `hasTail`.
     */
    unstable_onRenderTail?: (tail: ReactElement) => ReactElement;

    /**
     * Содержимое попапа
     */
    children?: ReactNode | ((props: { tailRef?: Ref<HTMLDivElement> }) => ReactNode);
}

export const cnPopup = cn('Popup2');

/**
 * Компонент для создания всплывающего окна (попапа).
 * @param {IPopupProps} props
 */
export const Popup: FC<IPopupProps> = ({
    addonAfter,
    addonBefore,
    children,
    className,
    direction,
    forceRender,
    hasTail,
    innerRef,
    keepMounted = true,
    position,
    scope = { current: canUseDOM() ? document.body : null },
    style,
    tailPosition,
    tailRef,
    tailSize,
    targetRef,
    visible,
    zIndex,
    unstable_onRenderTail,
    // Извлекаем свойства, т.к. они не нужны на DOM узле
    // FIXME: https://github.com/bem/bem-react/issues/381
    // @ts-ignore
    theme: _theme,
    // @ts-ignore
    view: _view,
    // @ts-ignore
    nonvisual: _nonvisual,
    ...props
}: IPopupProps) => {
    const [isFirstRender, forceUpdate] = useState(true);

    useUpdateEffect(() => {
        if (isFirstRender && visible) {
            forceUpdate(false);
        }
    }, []);

    useEffect(() => {
        if (isFirstRender && (forceRender || visible)) {
            forceUpdate(false);
        }
    }, []);

    const { Tail } = useComponentRegistry<IPopupRegistry>(cnPopup());

    if ((!visible && !keepMounted) || !canUseDOM() || scope.current === null || isFirstRender) {
        return null;
    }

    return createPortal(
        <div
            {...props}
            className={cnPopup({ visible, direction }, [className])}
            ref={mergeAllRefs(innerRef, targetRef)}
            style={{ ...style, ...position, zIndex }}
        >
            {addonBefore}
            {typeof children === 'function' ? children({ tailRef }) : children}
            {addonAfter}
            {unstable_onRenderTail &&
                unstable_onRenderTail(
                    <Tail innerRef={tailRef} style={{ ...tailPosition, height: tailSize, width: tailSize }} />,
                )}
            {!unstable_onRenderTail && hasTail && (
                <Tail innerRef={tailRef} style={{ ...tailPosition, height: tailSize, width: tailSize }} />
            )}
        </div>,
        scope.current,
    );
};

Popup.displayName = cnPopup();
