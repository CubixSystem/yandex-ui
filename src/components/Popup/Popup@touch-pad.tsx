import { withRegistry } from '@bem-react/di';

import '../../polyfills/click-event';
import { Popup as PopupCommon } from './Popup';
import { popupRegistry } from './Popup.registry/touch-pad';

export * from './Popup';
export const Popup = withRegistry(popupRegistry)(PopupCommon);
