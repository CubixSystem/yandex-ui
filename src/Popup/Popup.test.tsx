import React, { ComponentType, createRef } from 'react';
import { ReactWrapper, mount, render } from 'enzyme';
import { ExtractProps } from '@bem-react/core';

import { Popup as PopupCommon } from './Popup';
import { Popup as PopupDesktop } from './Popup@desktop';
import { Popup as PopupTouchPad } from './Popup@touch-phone';
import { Popup as PopupTouchPhone } from './Popup@touch-pad';
import { serverEnvironmentSetup, delay } from '../internal/utils';
import { Nullable } from '../typings/utility-types';

const platforms = [['desktop', PopupDesktop], ['touch-pad', PopupTouchPad], ['touch-phone', PopupTouchPhone]];

type PopupProps = ExtractProps<typeof PopupDesktop>;

describe.each<any>(platforms)('Popup@%s', (_platform, Popup: ComponentType<PopupProps>) => {
    let wrapper: Nullable<ReactWrapper> = null;

    afterEach(() => {
        if (wrapper !== null && wrapper.length > 0) {
            wrapper.unmount();
        }
    });

    describe('server environment', () => {
        serverEnvironmentSetup();

        test('должен вернуть полный шаблон компонента (snapshot)', () => {
            expect(
                render(
                    <Popup hasTail className="mix">
                        content
                    </Popup>,
                ),
            ).toMatchSnapshot();
        });
    });

    describe('client environment', () => {
        test('должен вернуть полный шаблон компонента (snapshot)', () => {
            wrapper = mount(
                <Popup visible hasTail className="mix">
                    content
                </Popup>,
            );
            expect(wrapper).toMatchSnapshot();
        });

        test('должен быть установлен корректный displayName', () => {
            expect(PopupCommon.displayName).toBe('Popup2');
        });

        test('должен устанавливать ссылку на корневой DOM элемент', async () => {
            const innerRef = createRef<HTMLDivElement>();
            wrapper = mount(<Popup visible innerRef={innerRef} />);
            await delay(100);
            expect(innerRef.current).not.toBe(null);
        });

        test('должен устанавливать ссылку на хвостик при наличии hasTail', async () => {
            const tailRef = createRef<HTMLDivElement>();
            wrapper = mount(<Popup visible tailRef={tailRef} hasTail />);
            await delay(100);
            expect(tailRef.current).not.toBeNull();
        });

        test('должен рендерить контент в addonAfter и addonBefore', () => {
            wrapper = mount(
                <Popup visible addonAfter={<i className="after" />} addonBefore={<i className="before" />} />,
            );
            expect(wrapper.find('.after')).toHaveLength(1);
            expect(wrapper.find('.before')).toHaveLength(1);
        });

        test('должен устанавливать position', () => {
            wrapper = mount(<Popup visible position={{ top: 10, left: 10 }} />);
            expect(wrapper.find('.Popup2').prop('style')).toEqual({ top: 10, left: 10 });
        });

        test('не должен рендериться в body если visible и keepMounted не установлены', async () => {
            const innerRef = createRef<HTMLDivElement>();
            wrapper = mount(<Popup innerRef={innerRef} keepMounted={false} />);
            await delay(100);
            expect(document.body.lastElementChild).toBeNull();
        });

        test('Должен удаляться из DOM если keepMounted не установлен', async () => {
            const innerRef = createRef<HTMLDivElement>();
            wrapper = mount(<Popup innerRef={innerRef} keepMounted={false} />);
            await delay(100);
            wrapper.setProps({ visible: true });
            expect(document.body.lastElementChild).toBe(innerRef.current);
            wrapper.setProps({ visible: false });
            expect(document.body.lastElementChild).toBeNull();
        });

        test('должен рендериться в body scope', async () => {
            const innerRef = createRef<HTMLDivElement>();
            wrapper = mount(<Popup innerRef={innerRef} visible />);
            await delay(100);
            expect(document.body.lastElementChild).toBe(innerRef.current);
        });

        test('должен рендериться в пользовательский scope', async () => {
            const element = document.createElement('div');
            document.body.appendChild(element);

            const innerRef = createRef<HTMLDivElement>();
            wrapper = mount(<Popup scope={{ current: element }} innerRef={innerRef} visible />);
            await delay(100);
            expect(element.lastElementChild).toBe(innerRef.current);
            document.body.removeChild(element);
        });

        test('должен устанавливать модификатор visible если открыт при инициализации', () => {
            wrapper = mount(<Popup visible />);
            expect(wrapper.find('.Popup2').hasClass('Popup2_visible')).toEqual(true);
        });

        test('должен устанавливать/удалять модификатор visible при открытии/закрытии', () => {
            wrapper = mount(<Popup />);
            wrapper.setProps({ visible: true });
            wrapper.update();
            expect(wrapper.find('.Popup2').hasClass('Popup2_visible')).toEqual(true);
            wrapper.setProps({ visible: false });
            expect(wrapper.find('.Popup2').hasClass('Popup2_visible')).toEqual(false);
        });

        test('должен лениво рендериться в DOM', () => {
            wrapper = mount(<Popup />);
            expect(wrapper.find('.Popup2').length).toBe(0);
            wrapper.setProps({ visible: true });
            wrapper.update();
            expect(wrapper.find('.Popup2').length).not.toBe(0);
        });
    });
});
