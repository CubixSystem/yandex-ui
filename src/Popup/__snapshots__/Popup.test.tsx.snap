// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup@desktop client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<RegistryResolver(Popup2)
  className="mix"
  hasTail={true}
  visible={true}
>
  <Popup2
    className="mix"
    hasTail={true}
    visible={true}
  >
    <Portal
      visible={true}
    >
      <Portal
        containerInfo={
          <body>
            <div
              class="Popup2 Popup2_visible mix"
            >
              content
              <div
                class="Popup2-Tail"
              >
                <svg
                  class="Popup2-TailSvg"
                  viewBox="0 0 24 24"
                >
                  <path
                    class="Popup2-TailFill"
                    d="M12 16C15 16 18 24 24 24H0C6 24 9 16 12 16Z"
                    fill="none"
                  />
                  <path
                    class="Popup2-TailBorder"
                    d="M8.05888 18.9807C6.77697 20.3966 5.32977 21.9951 3.57172 23H0C2.53449 23 4.52121 21.3153 6.39546 19.3161C6.68742 19.0046 6.98315 18.6782 7.27806 18.3526L7.27807 18.3526C7.87384 17.6949 8.46631 17.0408 9.01776 16.5218C9.82327 15.7637 10.827 15 12 15C13.173 15 14.1767 15.7637 14.9822 16.5218C15.5337 17.0408 16.1262 17.6949 16.7219 18.3526L16.7219 18.3526L16.7222 18.3529C17.017 18.6783 17.3127 19.0047 17.6045 19.3161C19.4788 21.3153 21.4655 23 24 23H20.4283C18.6702 21.9951 17.223 20.3966 15.9411 18.9807C14.4853 17.3726 13.2426 16 12 16C10.7574 16 9.51472 17.3726 8.05888 18.9807Z"
                    fill="none"
                  />
                </svg>
              </div>
            </div>
          </body>
        }
      >
        <div
          className="Popup2 Popup2_visible mix"
          style={
            Object {
              "zIndex": undefined,
            }
          }
        >
          content
          <ForwardRef
            style={
              Object {
                "height": undefined,
                "width": undefined,
              }
            }
          >
            <div
              className="Popup2-Tail"
              style={
                Object {
                  "height": undefined,
                  "width": undefined,
                }
              }
            >
              <svg
                className="Popup2-TailSvg"
                viewBox="0 0 24 24"
              >
                <path
                  className="Popup2-TailFill"
                  d="M12 16C15 16 18 24 24 24H0C6 24 9 16 12 16Z"
                  fill="none"
                />
                <path
                  className="Popup2-TailBorder"
                  d="M8.05888 18.9807C6.77697 20.3966 5.32977 21.9951 3.57172 23H0C2.53449 23 4.52121 21.3153 6.39546 19.3161C6.68742 19.0046 6.98315 18.6782 7.27806 18.3526L7.27807 18.3526C7.87384 17.6949 8.46631 17.0408 9.01776 16.5218C9.82327 15.7637 10.827 15 12 15C13.173 15 14.1767 15.7637 14.9822 16.5218C15.5337 17.0408 16.1262 17.6949 16.7219 18.3526L16.7219 18.3526L16.7222 18.3529C17.017 18.6783 17.3127 19.0047 17.6045 19.3161C19.4788 21.3153 21.4655 23 24 23H20.4283C18.6702 21.9951 17.223 20.3966 15.9411 18.9807C14.4853 17.3726 13.2426 16 12 16C10.7574 16 9.51472 17.3726 8.05888 18.9807Z"
                  fill="none"
                />
              </svg>
            </div>
          </ForwardRef>
        </div>
      </Portal>
    </Portal>
  </Popup2>
</RegistryResolver(Popup2)>
`;

exports[`Popup@desktop server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `null`;

exports[`Popup@touch-pad client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<RegistryResolver(Popup2)
  className="mix"
  hasTail={true}
  visible={true}
>
  <Popup2
    className="mix"
    hasTail={true}
    visible={true}
  >
    <Portal
      visible={true}
    >
      <Portal
        containerInfo={
          <body>
            <div
              class="Popup2 Popup2_visible mix"
            >
              content
              <div
                class="Popup2-Tail"
              >
                <svg
                  class="Popup2-TailSvg"
                  viewBox="0 0 24 24"
                >
                  <path
                    class="Popup2-TailFill"
                    d="M12 16C15 16 18 24 24 24H0C6 24 9 16 12 16Z"
                    fill="none"
                  />
                  <path
                    class="Popup2-TailBorder"
                    d="M8.05888 18.9807C6.77697 20.3966 5.32977 21.9951 3.57172 23H0C2.53449 23 4.52121 21.3153 6.39546 19.3161C6.68742 19.0046 6.98315 18.6782 7.27806 18.3526L7.27807 18.3526C7.87384 17.6949 8.46631 17.0408 9.01776 16.5218C9.82327 15.7637 10.827 15 12 15C13.173 15 14.1767 15.7637 14.9822 16.5218C15.5337 17.0408 16.1262 17.6949 16.7219 18.3526L16.7219 18.3526L16.7222 18.3529C17.017 18.6783 17.3127 19.0047 17.6045 19.3161C19.4788 21.3153 21.4655 23 24 23H20.4283C18.6702 21.9951 17.223 20.3966 15.9411 18.9807C14.4853 17.3726 13.2426 16 12 16C10.7574 16 9.51472 17.3726 8.05888 18.9807Z"
                    fill="none"
                  />
                </svg>
              </div>
            </div>
          </body>
        }
      >
        <div
          className="Popup2 Popup2_visible mix"
          style={
            Object {
              "zIndex": undefined,
            }
          }
        >
          content
          <ForwardRef
            style={
              Object {
                "height": undefined,
                "width": undefined,
              }
            }
          >
            <div
              className="Popup2-Tail"
              style={
                Object {
                  "height": undefined,
                  "width": undefined,
                }
              }
            >
              <svg
                className="Popup2-TailSvg"
                viewBox="0 0 24 24"
              >
                <path
                  className="Popup2-TailFill"
                  d="M12 16C15 16 18 24 24 24H0C6 24 9 16 12 16Z"
                  fill="none"
                />
                <path
                  className="Popup2-TailBorder"
                  d="M8.05888 18.9807C6.77697 20.3966 5.32977 21.9951 3.57172 23H0C2.53449 23 4.52121 21.3153 6.39546 19.3161C6.68742 19.0046 6.98315 18.6782 7.27806 18.3526L7.27807 18.3526C7.87384 17.6949 8.46631 17.0408 9.01776 16.5218C9.82327 15.7637 10.827 15 12 15C13.173 15 14.1767 15.7637 14.9822 16.5218C15.5337 17.0408 16.1262 17.6949 16.7219 18.3526L16.7219 18.3526L16.7222 18.3529C17.017 18.6783 17.3127 19.0047 17.6045 19.3161C19.4788 21.3153 21.4655 23 24 23H20.4283C18.6702 21.9951 17.223 20.3966 15.9411 18.9807C14.4853 17.3726 13.2426 16 12 16C10.7574 16 9.51472 17.3726 8.05888 18.9807Z"
                  fill="none"
                />
              </svg>
            </div>
          </ForwardRef>
        </div>
      </Portal>
    </Portal>
  </Popup2>
</RegistryResolver(Popup2)>
`;

exports[`Popup@touch-pad server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `null`;

exports[`Popup@touch-phone client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<RegistryResolver(Popup2)
  className="mix"
  hasTail={true}
  visible={true}
>
  <Popup2
    className="mix"
    hasTail={true}
    visible={true}
  >
    <Portal
      visible={true}
    >
      <Portal
        containerInfo={
          <body>
            <div
              class="Popup2 Popup2_visible mix"
            >
              content
              <div
                class="Popup2-Tail"
              >
                <svg
                  class="Popup2-TailSvg"
                  viewBox="0 0 24 24"
                >
                  <path
                    class="Popup2-TailFill"
                    d="M12 16C15 16 18 24 24 24H0C6 24 9 16 12 16Z"
                    fill="none"
                  />
                  <path
                    class="Popup2-TailBorder"
                    d="M8.05888 18.9807C6.77697 20.3966 5.32977 21.9951 3.57172 23H0C2.53449 23 4.52121 21.3153 6.39546 19.3161C6.68742 19.0046 6.98315 18.6782 7.27806 18.3526L7.27807 18.3526C7.87384 17.6949 8.46631 17.0408 9.01776 16.5218C9.82327 15.7637 10.827 15 12 15C13.173 15 14.1767 15.7637 14.9822 16.5218C15.5337 17.0408 16.1262 17.6949 16.7219 18.3526L16.7219 18.3526L16.7222 18.3529C17.017 18.6783 17.3127 19.0047 17.6045 19.3161C19.4788 21.3153 21.4655 23 24 23H20.4283C18.6702 21.9951 17.223 20.3966 15.9411 18.9807C14.4853 17.3726 13.2426 16 12 16C10.7574 16 9.51472 17.3726 8.05888 18.9807Z"
                    fill="none"
                  />
                </svg>
              </div>
            </div>
          </body>
        }
      >
        <div
          className="Popup2 Popup2_visible mix"
          style={
            Object {
              "zIndex": undefined,
            }
          }
        >
          content
          <ForwardRef
            style={
              Object {
                "height": undefined,
                "width": undefined,
              }
            }
          >
            <div
              className="Popup2-Tail"
              style={
                Object {
                  "height": undefined,
                  "width": undefined,
                }
              }
            >
              <svg
                className="Popup2-TailSvg"
                viewBox="0 0 24 24"
              >
                <path
                  className="Popup2-TailFill"
                  d="M12 16C15 16 18 24 24 24H0C6 24 9 16 12 16Z"
                  fill="none"
                />
                <path
                  className="Popup2-TailBorder"
                  d="M8.05888 18.9807C6.77697 20.3966 5.32977 21.9951 3.57172 23H0C2.53449 23 4.52121 21.3153 6.39546 19.3161C6.68742 19.0046 6.98315 18.6782 7.27806 18.3526L7.27807 18.3526C7.87384 17.6949 8.46631 17.0408 9.01776 16.5218C9.82327 15.7637 10.827 15 12 15C13.173 15 14.1767 15.7637 14.9822 16.5218C15.5337 17.0408 16.1262 17.6949 16.7219 18.3526L16.7219 18.3526L16.7222 18.3529C17.017 18.6783 17.3127 19.0047 17.6045 19.3161C19.4788 21.3153 21.4655 23 24 23H20.4283C18.6702 21.9951 17.223 20.3966 15.9411 18.9807C14.4853 17.3726 13.2426 16 12 16C10.7574 16 9.51472 17.3726 8.05888 18.9807Z"
                  fill="none"
                />
              </svg>
            </div>
          </ForwardRef>
        </div>
      </Portal>
    </Portal>
  </Popup2>
</RegistryResolver(Popup2)>
`;

exports[`Popup@touch-phone server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `null`;
